pkg/vm/memory/memory_value.go

```diff
@@ -259,7 +259,20 @@ func (mv *MemoryValue) Add(lhs, rhs *MemoryValue) error {
 // Subs two memory values if they're in the same segment or the rhs is a Felt.
 func (mv *MemoryValue) Sub(lhs, rhs *MemoryValue) error {
 	if lhs.IsAddress() {
-		return mv.addrUnsafe().Sub(lhs.addrUnsafe(), rhs.Any())
+		asAddr := mv.addrUnsafe()
+		if rhs.IsAddress() {
+			// The result is the offset value, felt-typed.
+			// See #284
+			mv.isFelt = true
+			mv.isAddress = false
+			if err := asAddr.Sub(lhs.addrUnsafe(), rhs.Any()); err != nil {
+				return err
+			}
+			mv.felt.SetUint64(asAddr.Offset)
+			return nil
+		}
+		// This code is executed for felt-typed rhs.
+		return asAddr.Sub(lhs.addrUnsafe(), rhs.Any())
```

Comment:
Shouldn't we then remove the `case` for rhs being a MemoryAddress in this function? or is still possible that this method is called from another place and that part gets executed anyway? Is it correct in that case to return an address instead of a felt?

Reply:
> Is it correct in that case to return an address instead of a felt?

It's hard for me to tell what are the original intents of the `MemoryAddress.Sub`.
I think it was intended to be a general purpose implementation of all supported overloadings for the addresses, but since it's implemented inside `MemoryAddress` and it does everything in-place, there is no way to make the operation produce `felt`.

If we would return the value instead, it could solve the problems, so I agree on that part. It looks like a better long-term solution. Let me try that right now.